apps/sertra/components/PackageDataFetcher.tsx

```diff
@@ -0,0 +1,65 @@
+import { useEffect, useState } from "react";
+import { ListingMod, TSListingMod } from "../api/models";
+import { ApiURLs, TsApiURLs } from "../api/urls";
+
+// TODO: To be replaced with ServerListingDetailData when the listing will contain an actual
+// object instead of an list of strings
+
+interface TempModRefType {
+  mods: string[];
+}
+
+export const FetchListingData = async (listingId: string) => {
+  const [listing, setListing] = useState<TempModRefType>();
+  useEffect(() => {
+    async function fetchListing() {
+      return await fetch(ApiURLs.ServerDetail(listingId));
+    }
+    async function jsonifyListing() {
+      setListing(await (await fetchListing()).json());
+    }
+    jsonifyListing();
+  });
+
+  const [mods, setMods] = useState<TSListingMod[]>([]);
+  useEffect(() => {
+    async function fetchMods() {
+      return await fetch(TsApiURLs.V1Packages("v-rising"));
+    }
+    async function jsonifyMods() {
+      setMods(await (await fetchMods()).json());
+    }
+    jsonifyMods();
+  });
+
+  const updated_listing_mods_data = [];
+  for (const mod_ref of listing.mods) {
+    const new_mod_data: ListingMod = {
+      name: "MOD DETAIL FETCH FAILED: " + mod_ref,
+      owner: null,
+      version: null,
+      icon_url: null,
+      description: null,
+    };
+    for (const mod of mods) {
+      if (mod_ref.startsWith(mod.full_name)) {
+        new_mod_data.name = mod.name;
+        new_mod_data.owner = mod.owner;
+        new_mod_data.version = null;
+        for (const version of mod.versions) {
+          if (mod_ref.endsWith(version.version_number)) {
```

Comment:
You **need** to do a full string comparison before you can be sure the mod actually matches, the current checks can let false positives through.

Likewise you can't assume the mod to match if just the start matches (on line 45), you can simply use that to reduce the amount of versions you enumerate over (just enumerate the mods with a matching start rather than _all_ versions). In other words you can't populate any data to the mod data before you have a full match, or otherwise you risk displaying invalid data.

I'd suggest you first flatten the entire mods list to a dictionary (you can use a [map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map) or a [record](https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type)), and then simply look up if a version exists there, rather than looping through everything multiple times.